adc/ade7913/ade7913-driver.c

```diff
@@ -0,0 +1,753 @@
+/*
+ * Phoenix-RTOS
+ *
+ * i.MX RT1170 ADE7913 test application
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Marcin Baran
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <errno.h>
+
+#include <sys/file.h>
+#include <sys/mman.h>
+#include <sys/msg.h>
+#include <sys/interrupt.h>
+#include <sys/platform.h>
+#include <sys/stat.h>
+#include <sys/threads.h>
+
+#include <imxrt-multi.h>
+#ifdef TARGET_IMXRT1170
+#include <phoenix/arch/imxrt1170.h>
+#endif
+
+#include <edma.h>
+
+#include "ade7913.h"
+#include "adc-api.h"
+
+#define COL_RED    "\033[1;31m"
+#define COL_CYAN   "\033[1;36m"
+#define COL_NORMAL "\033[0m"
+
+#define LOG_TAG "ade7913: "
+
+#ifdef NDEBUG
+#define log_debug(fmt, ...)
+#else
+#define log_debug(fmt, ...) \
+	do { \
+		printf(LOG_TAG fmt "\n", ##__VA_ARGS__); \
+	} while (0)
+#endif
+
+#define log_info(fmt, ...) \
+	do { \
+		printf(LOG_TAG COL_CYAN fmt COL_NORMAL "\n", ##__VA_ARGS__); \
+	} while (0)
+#define log_error(fmt, ...) \
+	do { \
+		printf(LOG_TAG COL_RED fmt COL_NORMAL "\n", ##__VA_ARGS__); \
+	} while (0)
+
+
+/*
+ * The buffer needs to be aligned with:
+ * num_of_devices * bytes_per_device * num_of_buffers
+ * so that in each filled buffer there will be same amount
+ * of samples for each device.
+ */
+#define ADC_BUFFER_SIZE             (6 * 16 * _PAGE_SIZE)
+#define NUM_OF_BUFFERS              (8)
+
+#define ADE7913_NUM_OF_BITS         24
+
+#define MAX_INIT_TRIES              (4)
+
+#define DREADY_DMA_CHANNEL          5
+#define GPIO_PIN_INTERRUPT          105 + 16
+#define SPI_RCV_DMA_CHANNEL         6
+#define SPI_RCV_DMA_REQUEST         36
+#define TCD_CSR_INTMAJOR_BIT        (1 << 1)
+#define TCD_CSR_ESG_BIT             (1 << 4)
+
+enum { gpio_dr = 0, gpio_gdir, gpio_psr, gpio_icr1, gpio_icr2, gpio_imr, gpio_isr, gpio_edge_sel };
+
+enum { spi_verid = 0, spi_param, spi_cr = 0x4, spi_sr, spi_ier, spi_der, spi_cfgr0, spi_cfgr1, spi_dmr0 = 0xc,
+	   spi_dmr1, spi_ccr = 0x10, spi_fcr = 0x16, spi_fsr, spi_tcr, spi_tdr, spi_rsr = 0x1c, spi_rdr };
+
+
+static uint32_t *gpio3_ptr = (uint32_t *)0x40134000;
+static volatile uint32_t *spi1_ptr = (uint32_t *)0x40114000;
+
+static const uint32_t adc_read_cmd_lookup[4] = { (0x4 << 24) | 0xFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF };
+
+static uint32_t spi_write_cmd_lookup[4] = { (spi_mode_3 << 30) | (1 << 27) | (0 << 24) | (spi_msb << 23) | (16 * 8 - 1),
+	(spi_mode_3 << 30) | (1 << 27) | (1 << 24) | (spi_msb << 23) | (16 * 8 - 1),
+	(spi_mode_3 << 30) | (1 << 27) | (2 << 24) | (spi_msb << 23) | (16 * 8 - 1),
+	(spi_mode_3 << 30) | (1 << 27) | (3 << 24) | (spi_msb << 23) | (16 * 8 - 1) };
+
+
+struct driver_common_s {
```

Comment:
No need to define `driver_common_s`, it's not used - this struct should be anonymous

Reply:
Done